src/java/org/apache/cassandra/nodes/Nodes.java

```diff
@@ -269,38 +273,54 @@ public IPeerInfo update(InetAddressAndPort peer, UnaryOperator<PeerInfo> update,
          */
         public IPeerInfo update(InetAddressAndPort peer, UnaryOperator<PeerInfo> update, boolean blocking, boolean force)
         {
-            return internalMap.compute(peer, (key, existingPeerInfo) -> {
-                PeerInfo updated = existingPeerInfo == null
-                                   ? update.apply(new PeerInfo().setPeerAddressAndPort(peer))
-                                   : update.apply(existingPeerInfo.duplicate()); // since we operate on mutable objects, we don't want to let the update function to operate on the live object
-
-                if (updated.getPeerAddressAndPort() == null)
-                    updated.setPeerAddressAndPort(peer);
-                else
-                    Preconditions.checkArgument(Objects.equals(updated.getPeerAddressAndPort(), peer));
-
-                updated.setRemoved(false);
-                save(existingPeerInfo, updated, blocking, force);
-                return updated;
-            });
+            try
+            {
+                return internalMap.compute(peer, (key, existingPeerInfo) -> {
+                    PeerInfo updated = existingPeerInfo == null
+                                       ? update.apply(new PeerInfo().setPeerAddressAndPort(peer))
+                                       : update.apply(existingPeerInfo.duplicate()); // since we operate on mutable objects, we don't want to let the update function to operate on the live object
+
+                    if (updated.getPeerAddressAndPort() == null)
+                        updated.setPeerAddressAndPort(peer);
+                    else
+                        Preconditions.checkArgument(Objects.equals(updated.getPeerAddressAndPort(), peer));
+
+                    updated.setRemoved(false);
+                    save(existingPeerInfo, updated, blocking, force);
+                    return updated;
+                });
+            }
+            finally
+            {
+                if (blocking)
```

Comment:
moving this out of internalMap.compute is the essential fix in this PR - sync does flushing, which may trigger compaction, which in turn may touch gossiper to modify the local node state, which goes back to Nodes and causes a deadlock. Since sync alone does not make any changes, we can safely move it out from the mapping function